docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,304 @@
+
+It is worth mentioning that the Infrastructure Machine Templates are not usually reconciled by themselves.
+Each infrastructure provider will be responsible for determining the best implementation for adding the
+status field based on the information available on their platform.
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU ResourceName = "cpu"
+    AutoscalerResourceMemory ResourceName = "memory"
+    AutoscalerResourceGPU ResourceName = "gpu"
+)
+
+// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineStatus struct {
+    Capacity ResourceList `json:"capacity,omitempty"`
+}
+
+// AutoscalerMachineTemplateStatus defines information that is used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineTemplateStatus struct {
+    Machine AutoscalerMachineStatus `json:"machine,omitempty"`
+}
+
+// DockerMachineTemplateStatus defines the observed state of a DockerMachineTemplate
+type DockerMachineTemplateStatus struct {
+    Autoscaler AutoscalerMachineTemplateStatus `json:"autoscaler,omitempty"`
+}
+
+// DockerMachineTemplate is the Schema for the dockermachinetemplates API.
+type DockerMachineTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
+    Status DockerMachineTemplateStatus `json:"status,omitempty"`
+}
+```
+_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/kubernetes/pkg/apis/core/types`_
+
+When used as a manifest, it would look like this:
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  template:
+    spec: {}
+status:
+  autoscaler:
+    machine:
+      capacity:
+        memory: 500mb
+        cpu: "1"
+        gpu: "1"
+```
+
+#### MachineSet and MachineDeployment Annotations
+
+In cases where a user needs to provide specific resource information for a
+MachineSet or MachineDeployment, or in cases where an infrastructure provider
+has not yet added the Infrastructure Machine Template status changes, they
+may use annotations to provide the information. The annotation values match
+API that is used in Infrastructure Machine Template, e.g. the memory and cpu
+annotations allow a `resource.Quantity` value and the gpu annotation allows the
+count of GPUs per instance.
+
+If a user wishes to specify the resource capacity through annotations, they
+may do by adding the following to any MachineSet or MachineDeployment (it is not required on both)
+that are participating in autoscaling:
+
+```
+kind: <MachineSet or MachineDeployment>
+metadata:
+  annotations:
+      autoscaler.cluster.x-k8s.io/machine-capacity-gpu: "1"
```

Comment:
Referring back to, can we make this example work for namespaced resources in https://github.com/kubernetes-sigs/cluster-api/pull/4283#issuecomment-901280727 as it will start to  become important in cloud providers that support different GPU architectures.

Reply:
Oh, ok. Is the GPU type going to be another annotation then?

Reply:
Yes, that's right 
![image](https://user-images.githubusercontent.com/1441070/134410647-1a4ff0de-fa99-4ae7-9e96-9fb95339b45f.png)

Reply:
> Is the GPU type going to be another annotation then?

i don't think we would need it as the simulator is able to determine the GPU type when it does the node creation simulation. (i am still understanding this code better, but i've seen it work)

> Slightly off-topic info which might be relevant: I'm not sure about annotations, but you can only have one / in labels, so this could also apply to annotations.

i might have misunderstood @randomvariable , but i was expecting we would allow this type of usage
```yaml
metadata:
  annotations:
      autoscaler.cluster.x-k8s.io/machine-capacity-gpu: "nvidia.com/gpu: 1"
```
but now that i type it out, it looks way too confusing.

Reply:
>  the simulator is able to determine the GPU type when it does the node creation simulation

Isn't this simply the case because there's provider specific code that maps instance types to GPU types? The AWS code also doesn't seem to know about AMD GPUs so presumably that's not supported at the moment. There seems to be a bit more logic for GKE if I'm reading the code correctly.

That said, I'm still trying to get my head around the cluster autoscaler space, so I'm probably missing something.

It's probably unlikely you'd have two different GPU types on the same host, so could probably do:

`autoscaler.cluster.x-k8s.io/machine-capacity-gpu: "1"`
`autoscaler.cluster.x-k8s.io/gpu-type: "nvidia.com/gpu"`


Which would work, but are we carrying baggage from the autoscaler's past here? The device plugin architecture is supposed to generic. In the on-premise space, we're starting to see nVidia suggest the appearance HPC workloads that use networking resources as well as GPUs, in particular Mellanox resources provided by Network Operator https://docs.mellanox.com/display/COKAN10/Network+Operator . In these cases we then start to care about RDMA & GPU resources, e.g. a pod with the declared requests:

``` yaml
apiVersion: v1
kind: Pod
metadata:
  name: testpod1
  annotations:
    k8s.v1.cni.cncf.io/networks: sriov-net
spec:
  containers:
  - name: appcntr1
    image: <image>
    imagePullPolicy: IfNotPresent
    securityContext:
      capabilities:
          add: ["IPC_LOCK"]
     volumeMounts:
      - mountPath: /dev/hugepages
        name: hugepage
      resources:
        requests:
          memory: 1Gi
          hugepages-1Gi: 2Gi
    command: [ "/bin/bash", "-c", "--" ]
    args: [ "while true; do sleep 300000; done;" ]
    resources:
      requests:
        nvidia.com/rdma_host_dev: '1'
        nvidia.com/gpu: '1'
      limits:
        nvidia.com/rdma_host_dev: '1'
        nvidia.com/gpu: '1'
  volumes:
   - name: hugepage
     emptyDir:
       medium: HugePages
```
This all seems fine from what has been proposed for the inframachinetemplate, since the definition of capacity can be a map. But this doesn't seem to work well for the manual annotations.

Reply:
i think the questions you raise are very poignant, and sadly i'm not sure i can answer them to the required depth necessary.

what i can say is that we currently run the autoscaling from zero on openshift with the annotations the way they exist and i've seen it do the proper thing when there are workload pending which have requested `nvidia.com/gpu: 1`.

> That said, I'm still trying to get my head around the cluster autoscaler space, so I'm probably missing something.

you and me both, i'm sure ;)

i am still diving through the simulator code to better understand how this works at the low level, but so far the best i can explain is that it appears that inside the simulator code is where it determines the exact type of GPU and if the pending pods will fit. my understanding is that it uses pieces of the core scheduler code to do this work, and that there is also a stage where it _predicts_ what the nodes will look like. it is during this prediction part that it find the appropriate resources. that's the best description i can give at the moment.

> It's probably unlikely you'd have two different GPU types on the same host, so could probably do:
>
> `autoscaler.cluster.x-k8s.io/machine-capacity-gpu: "1"`
> `autoscaler.cluster.x-k8s.io/gpu-type: "nvidia.com/gpu"`
>
> Which would work, but are we carrying baggage from the autoscaler's past here?

yeah, that could be possible. plus, i don't think adding that extra annotation would necessarily make any difference. my current observations are that this works with a simple numeric representation of the count of GPUs available. i'd prefer to keep this more simple for now and then add to it if we need to.

>  In the on-premise space, we're starting to see nVidia suggest the appearance HPC workloads that use networking resources as well as GPUs, in particular Mellanox resources provided by Network Operator https://docs.mellanox.com/display/COKAN10/Network+Operator . In these cases we then start to care about RDMA & GPU resources, e.g. a pod with the declared requests:

this is an excellent call out and something that we are seeing as well. i'm not sure if this is being handled just yet in the autoscaler, but i think our best option here is to keep consistent with what the autoscaler currently wants and help to make this better as we see the need. i don't have a ton of insight into the other accelerators just yet as i have been focused on the GPU stuff, but there are extension points within the autoscaler (eg custom node processors) which i imagine will help with this effort.

> This all seems fine from what has been proposed for the inframachinetemplate, since the definition of capacity can be a map. But this doesn't seem to work well for the manual annotations.

another great point, although my current understanding is that even with the expanded resource capacity status we are adding in this enhancement it will still boil down to a "number of GPUs present" at the API level during the initial node group pruning.

my preference is to stick with the annotations as they currently are, but i will add the `gpu-type` option if this is a blocker for you.

Reply:
@randomvariable i double checked the code and you are correct, we will need to have a gpu-type field as well. i'm going to update again.